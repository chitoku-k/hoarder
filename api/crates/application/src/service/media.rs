#[cfg_attr(feature = "test-mock", mockall::automock)]
pub trait MediaURLFactoryInterface: Send + Sync + 'static {
    fn public_url(&self, original_url: &str) -> Option<String>;
}
