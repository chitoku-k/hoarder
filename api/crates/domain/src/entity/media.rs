use chrono::{DateTime, Utc};
use derive_more::{Deref, Display, From};
use ordermap::OrderMap;
use serde::{Deserialize, Serialize};
use thiserror::Error;
use uuid::Uuid;

use crate::entity::{
    replicas::Replica,
    sources::Source,
    tag_types::TagType,
    tags::Tag,
};

#[derive(Clone, Copy, Debug, Default, Deref, Deserialize, Display, Eq, From, Hash, Ord, PartialEq, PartialOrd, Serialize)]
pub struct MediumId(Uuid);

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct Medium {
    pub id: MediumId,
    pub sources: Vec<Source>,
    pub tags: OrderMap<TagType, Vec<Tag>>,
    pub replicas: Vec<Replica>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Error)]
pub enum MediumError {
    #[error("medium not found: {0}")]
    NotFound(MediumId),
}
