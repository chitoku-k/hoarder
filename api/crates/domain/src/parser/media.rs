use std::future::Future;

use derive_more::{Constructor, From};

use crate::entity::replicas::Size;

#[cfg_attr(feature = "test-mock", mockall::automock)]
pub trait MediumImageParser: Send + Sync + 'static {
    /// Gets metadata for the image on the given path.
    fn get_metadata(&self, path: &str) -> impl Future<Output = anyhow::Result<MediumMetadata>> + Send;
}

#[derive(Clone, Constructor, Copy, Debug, Eq, From, PartialEq)]
pub struct MediumMetadata {
    mime_type: &'static str,
    width: u32,
    height: u32,
}

impl MediumMetadata {
    pub const fn mime_type(&self) -> &'static str {
        self.mime_type
    }

    pub fn size(&self) -> Size {
        Size::new(self.width, self.height)
    }
}
