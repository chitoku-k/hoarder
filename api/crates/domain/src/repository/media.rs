use std::future::Future;

use chrono::{DateTime, Utc};

use crate::{
    entity::{
        media::{Medium, MediumId},
        replicas::ReplicaId,
        sources::SourceId,
        tag_types::TagTypeId,
        tags::{TagDepth, TagId},
    },
    error::Result,
    repository::{DeleteResult, Direction, Order},
};

#[cfg_attr(feature = "test-mock", mockall::automock)]
pub trait MediaRepository: Send + Sync + 'static {
    /// Creates a medium.
    fn create<T, U>(&self, source_ids: T, created_at: Option<DateTime<Utc>>, tag_tag_type_ids: U, tag_depth: Option<TagDepth>, sources: bool) -> impl Future<Output = Result<Medium>> + Send
    where
        T: IntoIterator<Item = SourceId> + Send + Sync + 'static,
        U: IntoIterator<Item = (TagId, TagTypeId)> + Send + Sync + 'static;

    /// Fetches media by IDs.
    fn fetch_by_ids<T>(&self, ids: T, tag_depth: Option<TagDepth>, replicas: bool, sources: bool) -> impl Future<Output = Result<Vec<Medium>>> + Send
    where
        T: IntoIterator<Item = MediumId> + Send + Sync + 'static;

    /// Fetches media by their associated source IDs.
    fn fetch_by_source_ids<T>(
        &self,
        source_ids: T,
        tag_depth: Option<TagDepth>,
        replicas: bool,
        sources: bool,
        cursor: Option<(DateTime<Utc>, MediumId)>,
        order: Order,
        direction: Direction,
        limit: u64,
    ) -> impl Future<Output = Result<Vec<Medium>>> + Send
    where
        T: IntoIterator<Item = SourceId> + Send + Sync + 'static;

    /// Fetches media by their associated tag IDs.
    fn fetch_by_tag_ids<T>(
        &self,
        tag_tag_type_ids: T,
        tag_depth: Option<TagDepth>,
        replicas: bool,
        sources: bool,
        cursor: Option<(DateTime<Utc>, MediumId)>,
        order: Order,
        direction: Direction,
        limit: u64,
    ) -> impl Future<Output = Result<Vec<Medium>>> + Send
    where
        T: IntoIterator<Item = (TagId, TagTypeId)> + Send + Sync + 'static;

    /// Fetches all media.
    fn fetch_all(
        &self,
        tag_depth: Option<TagDepth>,
        replicas: bool,
        sources: bool,
        cursor: Option<(DateTime<Utc>, MediumId)>,
        order: Order,
        direction: Direction,
        limit: u64,
    ) -> impl Future<Output = Result<Vec<Medium>>> + Send;

    /// Updates the medium by ID.
    fn update_by_id<T, U, V, W, X>(
        &self,
        id: MediumId,
        add_source_ids: T,
        remove_source_ids: U,
        add_tag_tag_type_ids: V,
        remove_tag_tag_type_ids: W,
        replica_orders: X,
        created_at: Option<DateTime<Utc>>,
        tag_depth: Option<TagDepth>,
        replicas: bool,
        sources: bool,
    ) -> impl Future<Output = Result<Medium>> + Send
    where
        T: IntoIterator<Item = SourceId> + Send + Sync + 'static,
        U: IntoIterator<Item = SourceId> + Send + Sync + 'static,
        V: IntoIterator<Item = (TagId, TagTypeId)> + Send + Sync + 'static,
        W: IntoIterator<Item = (TagId, TagTypeId)> + Send + Sync + 'static,
        X: IntoIterator<Item = ReplicaId> + Send + Sync + 'static;

    /// Deletes the medium by ID.
    fn delete_by_id(&self, id: MediumId) -> impl Future<Output = Result<DeleteResult>> + Send;
}
