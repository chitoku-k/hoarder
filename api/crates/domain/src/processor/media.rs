use std::{future::Future, io::{BufRead, Seek}};

use crate::entity::replicas::{OriginalImage, ThumbnailImage};

#[cfg_attr(feature = "test-mock", mockall::automock)]
pub trait MediumImageProcessor: Send + Sync + 'static {
    /// Generates a thumbnail for image on the given path.
    fn generate_thumbnail<R>(&self, read: R) -> impl Future<Output = anyhow::Result<(OriginalImage, ThumbnailImage)>> + Send
    where
        R: BufRead + Seek + Send + 'static;
}
