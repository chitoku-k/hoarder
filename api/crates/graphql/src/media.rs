use async_graphql::{
    connection::{CursorType, DefaultEdgeName, Edge, EmptyFields},
    SimpleObject,
};
use base64::prelude::{BASE64_STANDARD, Engine};
use chrono::{DateTime, Utc};
use derive_more::Constructor;
use domain::entity::media::{self, MediumId};
use uuid::Uuid;

use crate::{
    error::ErrorKind,
    replicas::Replica,
    sources::Source,
    tags::TagTagType,
};

#[derive(SimpleObject)]
pub(crate) struct Medium {
    id: Uuid,
    sources: Vec<Source>,
    tags: Vec<TagTagType>,
    replicas: Vec<Replica>,
    created_at: DateTime<Utc>,
    updated_at: DateTime<Utc>,
}

#[derive(Clone, Constructor, Debug, Eq, PartialEq)]
pub(crate) struct MediumCursor(DateTime<Utc>, Uuid);

impl TryFrom<media::Medium> for Medium {
    type Error = ErrorKind;

    fn try_from(medium: media::Medium) -> Result<Self, Self::Error> {
        let sources: Result<_, _> = medium.sources
            .into_iter()
            .map(TryInto::try_into)
            .collect();

        let tags = medium.tags
            .into_iter()
            .flat_map(|(tag_type, tags)| {
                tags
                    .into_iter()
                    .map(move |tag| TagTagType::new(tag.into(), tag_type.clone().into()))
            })
            .collect();

        let replicas = medium.replicas
            .into_iter()
            .map(Into::into)
            .collect();

        Ok(Self {
            id: *medium.id,
            sources: sources?,
            tags,
            replicas,
            created_at: medium.created_at,
            updated_at: medium.updated_at,
        })
    }
}

impl MediumCursor {
    const DELIMITER: char = '\x00';

    pub fn into_inner(self) -> (DateTime<Utc>, MediumId) {
        (self.0, self.1.into())
    }
}

impl CursorType for MediumCursor {
    type Error = ErrorKind;

    fn decode_cursor(s: &str) -> Result<Self, Self::Error> {
        let bin = BASE64_STANDARD.decode(s).map_err(|_| ErrorKind::CursorInvalid)?;
        let str = String::from_utf8(bin).map_err(|_| ErrorKind::CursorInvalid)?;

        let (datetime, uuid) = str.split_once(Self::DELIMITER).ok_or(ErrorKind::CursorInvalid)?;
        let datetime = DateTime::parse_from_rfc3339(datetime).map_err(|_| ErrorKind::CursorInvalid)?.into();
        let uuid = Uuid::parse_str(uuid).map_err(|_| ErrorKind::CursorInvalid)?;

        Ok(MediumCursor::new(datetime, uuid))
    }

    fn encode_cursor(&self) -> String {
        let datetime = self.0.to_rfc3339();
        let uuid = self.1;
        let str = format!("{}{}{}", datetime, Self::DELIMITER, uuid);

        BASE64_STANDARD.encode(str)
    }
}

impl From<Medium> for Edge<MediumCursor, Medium, EmptyFields, DefaultEdgeName> {
    fn from(medium: Medium) -> Self {
        let cursor = MediumCursor::new(medium.created_at, medium.id);
        Edge::new(cursor, medium)
    }
}

#[cfg(test)]
mod tests {
    use chrono::TimeZone;
    use pretty_assertions::assert_eq;
    use uuid::uuid;

    use super::*;

    #[test]
    fn medium_cursor_into_inner() {
        let cursor = MediumCursor::new(
            Utc.with_ymd_and_hms(2022, 1, 1, 3, 4, 15).unwrap(),
            uuid!("6356503d-6ab6-4e39-bb86-3311219c7fd1"),
        );
        let actual = cursor.into_inner();

        assert_eq!(
            actual,
            (
                Utc.with_ymd_and_hms(2022, 1, 1, 3, 4, 15).unwrap(),
                MediumId::from(uuid!("6356503d-6ab6-4e39-bb86-3311219c7fd1")),
            ),
        );
    }

    #[test]
    fn media_cursor_encode_cursor_succeeds() {
        let cursor = MediumCursor::new(
            Utc.with_ymd_and_hms(2022, 1, 1, 3, 4, 15).unwrap(),
            uuid!("6356503d-6ab6-4e39-bb86-3311219c7fd1"),
        );
        let actual = cursor.encode_cursor();

        assert_eq!(actual, "MjAyMi0wMS0wMVQwMzowNDoxNSswMDowMAA2MzU2NTAzZC02YWI2LTRlMzktYmI4Ni0zMzExMjE5YzdmZDE=".to_string());
    }

    #[test]
    fn media_cursor_decode_cursor_succeeds() {
        let actual = MediumCursor::decode_cursor("MjAyMi0wMS0wMVQwMzowNDoxNSswMDowMAA2MzU2NTAzZC02YWI2LTRlMzktYmI4Ni0zMzExMjE5YzdmZDE=").unwrap();

        assert_eq!(
            actual,
            MediumCursor::new(
                Utc.with_ymd_and_hms(2022, 1, 1, 3, 4, 15).unwrap(),
                uuid!("6356503d-6ab6-4e39-bb86-3311219c7fd1"),
            ),
        );
    }

    #[test]
    fn media_cursor_decode_cursor_fails() {
        let actual = MediumCursor::decode_cursor("====");

        assert!(actual.is_err());
    }
}
