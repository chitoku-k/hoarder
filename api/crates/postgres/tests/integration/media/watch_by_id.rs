use chrono::{TimeZone, Utc};
use domain::{
    entity::{
        media::{Medium, MediumId},
        replicas::{Replica, ReplicaId, ReplicaStatus, Size},
    },
    repository::media::MediaRepository,
};
use futures::{pin_mut, TryStreamExt};
use ordermap::OrderMap;
use postgres::media::PostgresMediaRepository;
use pretty_assertions::assert_eq;
use test_context::test_context;
use uuid::uuid;

use super::DatabaseContext;

#[test_context(DatabaseContext)]
#[tokio::test]
async fn succeeds(ctx: &DatabaseContext) {
    let repository = PostgresMediaRepository::new(ctx.pool.clone());
    let stream = repository.watch_by_id(
        MediumId::from(uuid!("2872ed9d-4db9-4b25-b86f-791ad009cc0a")),
        None,
        true,
        false,
    ).await.unwrap();
    pin_mut!(stream);

    let actual = stream.try_next().await.unwrap();

    assert_eq!(actual, Some(Medium {
        id: MediumId::from(uuid!("2872ed9d-4db9-4b25-b86f-791ad009cc0a")),
        sources: Vec::new(),
        tags: OrderMap::new(),
        replicas: vec![
            Replica {
                id: ReplicaId::from(uuid!("b7a54e0b-6ab3-4385-a18b-bacadff6b18d")),
                display_order: 1,
                thumbnail: None,
                original_url: "file:///b7a54e0b-6ab3-4385-a18b-bacadff6b18d.jpg".to_string(),
                mime_type: Some("image/jpeg".to_string()),
                size: Some(Size::new(1800, 2400)),
                status: ReplicaStatus::Ready,
                created_at: Utc.with_ymd_and_hms(2022, 1, 2, 3, 4, 6).unwrap(),
                updated_at: Utc.with_ymd_and_hms(2022, 2, 3, 4, 5, 9).unwrap(),
            },
            Replica {
                id: ReplicaId::from(uuid!("790dc278-2c53-4988-883c-43a037664b24")),
                display_order: 2,
                thumbnail: None,
                original_url: "file:///790dc278-2c53-4988-883c-43a037664b24.jpg".to_string(),
                mime_type: Some("image/jpeg".to_string()),
                size: Some(Size::new(1800, 2400)),
                status: ReplicaStatus::Ready,
                created_at: Utc.with_ymd_and_hms(2022, 1, 2, 3, 4, 6).unwrap(),
                updated_at: Utc.with_ymd_and_hms(2022, 2, 3, 4, 5, 6).unwrap(),
            },
        ],
        created_at: Utc.with_ymd_and_hms(2022, 1, 2, 3, 4, 6).unwrap(),
        updated_at: Utc.with_ymd_and_hms(2022, 2, 3, 4, 5, 9).unwrap(),
    }));

    sqlx::query(r#"UPDATE "replicas" SET "mime_type" = $1, "width" = $2, "height" = $3, "phase" = $4 WHERE "id" = $5"#)
        .bind(None::<&str>)
        .bind(None::<i32>)
        .bind(None::<i32>)
        .bind("processing")
        .bind(uuid!("b7a54e0b-6ab3-4385-a18b-bacadff6b18d"))
        .execute(&ctx.pool)
        .await
        .unwrap();

    sqlx::query(r#"SELECT pg_notify($1, $2)"#)
        .bind("replicas")
        .bind(r#"{"id":"b7a54e0b-6ab3-4385-a18b-bacadff6b18d","medium_id":"2872ed9d-4db9-4b25-b86f-791ad009cc0a"}"#)
        .execute(&ctx.pool)
        .await
        .unwrap();

    let actual = stream.try_next().await.unwrap();

    assert_eq!(actual, Some(Medium {
        id: MediumId::from(uuid!("2872ed9d-4db9-4b25-b86f-791ad009cc0a")),
        sources: Vec::new(),
        tags: OrderMap::new(),
        replicas: vec![
            Replica {
                id: ReplicaId::from(uuid!("b7a54e0b-6ab3-4385-a18b-bacadff6b18d")),
                display_order: 1,
                thumbnail: None,
                original_url: "file:///b7a54e0b-6ab3-4385-a18b-bacadff6b18d.jpg".to_string(),
                mime_type: None,
                size: None,
                status: ReplicaStatus::Processing,
                created_at: Utc.with_ymd_and_hms(2022, 1, 2, 3, 4, 6).unwrap(),
                updated_at: Utc.with_ymd_and_hms(2022, 2, 3, 4, 5, 9).unwrap(),
            },
            Replica {
                id: ReplicaId::from(uuid!("790dc278-2c53-4988-883c-43a037664b24")),
                display_order: 2,
                thumbnail: None,
                original_url: "file:///790dc278-2c53-4988-883c-43a037664b24.jpg".to_string(),
                mime_type: Some("image/jpeg".to_string()),
                size: Some(Size::new(1800, 2400)),
                status: ReplicaStatus::Ready,
                created_at: Utc.with_ymd_and_hms(2022, 1, 2, 3, 4, 6).unwrap(),
                updated_at: Utc.with_ymd_and_hms(2022, 2, 3, 4, 5, 6).unwrap(),
            },
        ],
        created_at: Utc.with_ymd_and_hms(2022, 1, 2, 3, 4, 6).unwrap(),
        updated_at: Utc.with_ymd_and_hms(2022, 2, 3, 4, 5, 9).unwrap(),
    }));
}
