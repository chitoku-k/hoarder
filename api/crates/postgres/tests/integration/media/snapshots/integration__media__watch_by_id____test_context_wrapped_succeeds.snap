---
source: crates/postgres/tests/integration/media/watch_by_id.rs
expression: ctx.queries()
---
[[queries]]
sql = 'LISTEN "replicas"'
rows_affected = 0
rows_returned = 0

[[queries]]
sql = '''
SELECT
  "id",
  "created_at",
  "updated_at"
FROM
  "media"
WHERE
  "id" = $1'''
rows_affected = 1
rows_returned = 1

[[queries]]
sql = '''
SELECT
  "replicas"."id" AS "replica_id",
  "replicas"."medium_id" AS "replica_medium_id",
  "replicas"."display_order" AS "replica_display_order",
  "replicas"."original_url" AS "replica_original_url",
  "replicas"."mime_type" AS "replica_mime_type",
  "replicas"."width" AS "replica_width",
  "replicas"."height" AS "replica_height",
  "replicas"."phase" AS "replica_phase",
  "replicas"."created_at" AS "replica_created_at",
  "replicas"."updated_at" AS "replica_updated_at",
  "thumbnails"."id" AS "thumbnail_id",
  "thumbnails"."width" AS "thumbnail_width",
  "thumbnails"."height" AS "thumbnail_height",
  "thumbnails"."created_at" AS "thumbnail_created_at",
  "thumbnails"."updated_at" AS "thumbnail_updated_at"
FROM
  "replicas"
LEFT JOIN
  "thumbnails" ON "replicas"."id" = "thumbnails"."replica_id"
WHERE
  "medium_id" IN ($1)
ORDER BY
  "medium_id" ASC,
  "display_order" ASC'''
rows_affected = 2
rows_returned = 2

[[queries]]
sql = '''
SELECT
  "id",
  "created_at",
  "updated_at"
FROM
  "media"
WHERE
  "id" = $1'''
rows_affected = 1
rows_returned = 1

[[queries]]
sql = '''
SELECT
  "replicas"."id" AS "replica_id",
  "replicas"."medium_id" AS "replica_medium_id",
  "replicas"."display_order" AS "replica_display_order",
  "replicas"."original_url" AS "replica_original_url",
  "replicas"."mime_type" AS "replica_mime_type",
  "replicas"."width" AS "replica_width",
  "replicas"."height" AS "replica_height",
  "replicas"."phase" AS "replica_phase",
  "replicas"."created_at" AS "replica_created_at",
  "replicas"."updated_at" AS "replica_updated_at",
  "thumbnails"."id" AS "thumbnail_id",
  "thumbnails"."width" AS "thumbnail_width",
  "thumbnails"."height" AS "thumbnail_height",
  "thumbnails"."created_at" AS "thumbnail_created_at",
  "thumbnails"."updated_at" AS "thumbnail_updated_at"
FROM
  "replicas"
LEFT JOIN
  "thumbnails" ON "replicas"."id" = "thumbnails"."replica_id"
WHERE
  "medium_id" IN ($1)
ORDER BY
  "medium_id" ASC,
  "display_order" ASC'''
rows_affected = 2
rows_returned = 2

[[queries]]
sql = '''
SELECT
  "id",
  "created_at",
  "updated_at"
FROM
  "media"
WHERE
  "id" = $1'''
rows_affected = 1
rows_returned = 1

[[queries]]
sql = '''
SELECT
  "replicas"."id" AS "replica_id",
  "replicas"."medium_id" AS "replica_medium_id",
  "replicas"."display_order" AS "replica_display_order",
  "replicas"."original_url" AS "replica_original_url",
  "replicas"."mime_type" AS "replica_mime_type",
  "replicas"."width" AS "replica_width",
  "replicas"."height" AS "replica_height",
  "replicas"."phase" AS "replica_phase",
  "replicas"."created_at" AS "replica_created_at",
  "replicas"."updated_at" AS "replica_updated_at",
  "thumbnails"."id" AS "thumbnail_id",
  "thumbnails"."width" AS "thumbnail_width",
  "thumbnails"."height" AS "thumbnail_height",
  "thumbnails"."created_at" AS "thumbnail_created_at",
  "thumbnails"."updated_at" AS "thumbnail_updated_at"
FROM
  "replicas"
LEFT JOIN
  "thumbnails" ON "replicas"."id" = "thumbnails"."replica_id"
WHERE
  "medium_id" IN ($1)
ORDER BY
  "medium_id" ASC,
  "display_order" ASC'''
rows_affected = 2
rows_returned = 2
